src/runtime/hexagon_remote/halide_remote.cpp

```diff
@@ -44,16 +87,227 @@ int halide_do_task(void *user_context, halide_task_t f, int idx,
                    uint8_t *closure) {
     return f(user_context, idx, closure);
 }
+void lock_work_queue() {
+    qurt_mutex_lock(&(work_queue.work_mutex));
+}
+void unlock_work_queue() {
+    qurt_mutex_unlock(&(work_queue.work_mutex));
+}
 
-int halide_do_par_for(void *user_context, halide_task_t f,
-                      int min, int size, uint8_t *closure) {
-    for (int x = min; x < min + size; x++) {
-        int result = halide_do_task(user_context, f, x, closure);
+// This function does the real work of the thread pool.
+// owned_job is used to tell the differece between the master thread
+// and the worker threads. If owned_job is non-null then it means that
+// we are in the master thread, i.e. the thread that put owned_job in
+// the work queue. This thread, like other worker threads in this function,
+// tries to acquire a lock on the work queue. Each thread then looks for
+// a job to do. In the case of the master thread, it is "owned_job" while
+// a worker thread has to iterate backwards over the work queue and find
+// a job to work on. Once a job has been found, the thread, after some
+// book-keeping, releases the lock and call halide_do_task to do the job.
+// After the job is done, the lock on the queue is re-acquired to update
+// the status and the thread loops again to look for new work. If a worker
+// thread doesn't find any new work, it goes to sleep until awoken by
+// the master thread.
+void goto_work(work *owned_job) {
+    work myjob;
+    // You can work only if you get a lock on the work queue.
+    unsigned int tid = qurt_thread_get_id();
+    FARF(LOW, "HVX_TP: %d: goto_work: Trying to get a lock on the work queue\n", tid);
+
+    // ***********************
+    // *** Lock work queue ***
+    // ***********************
+    lock_work_queue();
+    FARF(LOW, "HVX_TP: %d: goto_work: got a lock on the work queue\n", tid);
+
+    // If I'm a job owner, then I was the thread that called
+    // do_par_for, and I should only stay in this function while there
+    // are active workers on the job (i.e. running()) or if the job
+    // hasn't been claimed entirely (!claimed()). If I'm a lowly worker
+    // thread, I should stay in this function as long as the work queue is running.
+    while (owned_job != NULL ? (owned_job->running() || !owned_job->claimed())
+           : work_queue.running()) {
+
+        FARF(LOW, "HVX_TP: %d: goto_work: In the main goto_work loop \n", tid);
+        work *jobs = work_queue.jobs;
+        work *job;
```

Comment:
I'd prefer if this was null initialized, so if some crazy bug happens/gets introduced, the access below the if isn't junk (could also add an assert below the if that job != null)